editor/src/clj/editor/code/view.clj

```diff
@@ -3360,20 +3513,25 @@
       (let [row (data/last-visible-row (:layout canvas-repaint-info))]
         (repaint-canvas! canvas-repaint-info (get-valid-syntax-info resource-node canvas-repaint-info row))))
 
-    ;; Show completion suggestions if appropriate.
     (when editable
-      (let [renderer (or (g/user-data view-node :completion-popup-renderer)
-                         (g/user-data!
-                           view-node
-                           :completion-popup-renderer
-                           (fx/create-renderer
-                             :error-handler error-reporting/report-exception!
-                             :middleware (comp
-                                           fxui/wrap-dedupe-desc
-                                           (fx/wrap-map-desc #'completion-popup-view))
-                             :opts {:fx.opt/map-event-handler #(handle-completion-popup-event view-node %)})))
-            ^Canvas canvas (:canvas canvas-repaint-info)]
-        (g/with-auto-evaluation-context evaluation-context
+      (g/with-auto-evaluation-context evaluation-context
+        ;; Show rename popup if appropriate
+        (advance-user-data-component!
+          view-node :rename-popup
+          (when (g/node-value view-node :rename-cursor-range evaluation-context)
+            (g/node-value view-node :rename-view evaluation-context)))
```

Comment:
This is in the `repaint-view!` function: we simply pull a cached component and advance the cljfx state. This is more performant/memory efficient if we compare it to the completion popup refresh below, which creates a new map on every repaint.